ocf/ocf_test.go

```diff
@@ -967,6 +968,200 @@ func TestEncoder_WriteHeaderError(t *testing.T) {
 	assert.Error(t, err)
 }
 
+func TestWithSchemaCache(t *testing.T) {
+	schema := `{
+		"type": "record",
+		"name": "Foo",
+		"namespace": "foo.bar.baz",
+		"fields": [
+			{
+				"name": "name",
+				"type": "string"
+			},
+			{
+				"name": "id",
+				"type": "long"
+			},
+			{
+				"name": "meta",
+				"type": {
+					"type": "array",
+					"items": {
+						"type": "record",
+						"name": "FooMetadataEntry",
+						"namespace": "foo.bar.baz",
+						"fields": [
+							{
+								"name": "key",
+								"type": "string"
+							},
+							{
+								"name": "values",
+								"type": {
+									"type": "array",
+									"items": "string"
+								}
+							}
+						]
+					}
+				}
+			}
+		]
+	}`
+	type metaEntry struct {
+		Key    string   `avro:"key"`
+		Values []string `avro:"values"`
+	}
+	type foo struct {
+		Name string      `avro:"name"`
+		ID   int64       `avro:"id"`
+		Meta []metaEntry `avro:"meta"`
+	}
+	encoderCache := &avro.SchemaCache{}
+	var buf bytes.Buffer
+	enc, err := ocf.NewEncoder(schema, &buf, ocf.WithEncoderSchemaCache(encoderCache))
+	require.NoError(t, err)
+	val := foo{
+		Name: "Bob Loblaw",
+		ID:   42,
+		Meta: []metaEntry{
+			{
+				Key:    "abc",
+				Values: []string{"123", "456"},
+			},
+		},
+	}
+	require.NoError(t, enc.Encode(val))
+	require.NoError(t, enc.Close())
+
+	assert.NotNil(t, encoderCache.Get("foo.bar.baz.Foo"))
+	assert.NotNil(t, encoderCache.Get("foo.bar.baz.FooMetadataEntry"))
+	assert.Nil(t, avro.DefaultSchemaCache.Get("foo.bar.baz.Foo"))
+	assert.Nil(t, avro.DefaultSchemaCache.Get("foo.bar.baz.FooMetadataEntry"))
+
+	decoderCache := &avro.SchemaCache{}
+	dec, err := ocf.NewDecoder(&buf, ocf.WithDecoderSchemaCache(decoderCache))
+	require.NoError(t, err)
+	require.True(t, dec.HasNext())
+	var roundTripVal foo
+	require.NoError(t, dec.Decode(&roundTripVal))
+	require.False(t, dec.HasNext())
+	require.Equal(t, val, roundTripVal)
+
+	assert.NotNil(t, decoderCache.Get("foo.bar.baz.Foo"))
+	assert.NotNil(t, decoderCache.Get("foo.bar.baz.FooMetadataEntry"))
+	assert.Nil(t, avro.DefaultSchemaCache.Get("foo.bar.baz.Foo"))
+	assert.Nil(t, avro.DefaultSchemaCache.Get("foo.bar.baz.FooMetadataEntry"))
+}
+
+func TestWithSchemaMarshaler(t *testing.T) {
```

Comment:
Would not not be simpler for this test to use a "Golden" file?

Reply:
Could you elaborate? Do you have examples/patterns in the repo already for this? Are you thinking of putting the schema JSON into a golden file? I'm used to that phrase being for validation of output (by comparing output to that of a golden file), but it wasn't clear that applies here. Unless you just wanted to have it marshal `dec.Schema()` and then compare that to a golden file (instead of just extracting the extra properties and checking them).

Reply:
Yes, that was my thought, to compare against a golden file to not extract the properties. This is done in `gen` already for examples.

Reply:
Done